File: iOS/LeetCodeSwift/LeetCodeSwift/Leetcode/Climbing_stairs.swift
//
//  Climbing_stairs.swift
//  LeetCodeSwift
//
//  Created by tree on 2018/3/12.
//  Copyright © 2018年 treee. All rights reserved.
//

import Foundation

extension Solution {
    /**
     You are climbing a stair case. It takes n steps to reach to the top.
     
     Each time you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top?
     
     Note: Given n will be a positive integer.
     
     当前的可能性为前两级的可能性之和
    */
    func climbStairs(_ n: Int) -> Int {
        if n < 3 { return max(n, 0) }
        var steps: [Int] = [0, 1, 2]
        for index in 3...n {
            steps.append(steps[index-1] + steps[index-2])
            print("第\(index)个\(steps)")
        }
        return steps.last ?? 0
    }
}

File: python/LeetCode/Climbing_Stairs.py
"""
爬楼梯

有一阶梯，需要n步爬上去，每步只能跨一或者二级，有几种方法可以到顶上

e.g:
Input: 2
Output:  2
1. 1步+1步
2. 2步


Input: 3
Output:  3
1. 1步+1步+1步
2. 2步+1步
3. 1步+2步

思路：
在大于2级的前提下，到达当前阶梯的可能的方式有上一级与上上一级的和
"""

class Solution(object):

    def climbingStairs(self, n: int) -> int:
        if n <= 3:
            return n
        steps = [0, 1, 2]
        for index in range(3,n+1):
            steps.append(steps[index-1] + steps[index-2])
        return steps[-1] or 0


if __name__ == '__main__':
    s = Solution()
    result = s.climbingStairs(5)
    print(result)